### ISSUE TITLE
[CSA] False negative: allocations reachable only through a destroyed owner are not reported

### ISSUE THREAD

**Issue by chisa22** [llvm/llvm-project#214226]
Hi, I found a family of false negatives in `cplusplus.NewDeleteLeaks`.

## Program

```cpp
struct Owner {
  int *member;
  Owner() : member(new int(42)) {}
};

void test() {
  Owner *owner = new Owner;
  delete owner;
}
```

`Owner` has no destructor that releases `member`. Deleting `owner` therefore makes the second allocation unreachable, but CSA emits no `cplusplus.NewDeleteLeaks` diagnostic.

The same behavior occurs for a linked-list child, pointer members inherited by a derived object, an omitted member in an explicit destructor, and objects stored in a dynamically allocated pointer array before only that array is deleted.

## CSA reproduction

```sh
clang++ --analyze \
  -Xclang -analyzer-checker=cplusplus.NewDeleteLeaks \
  owned_pointer_member_leak.cpp
```

## Comparison with related programs

A direct discarded allocation is reported:

```cpp
void direct_leak() {
  new int(42); // warning: potential memory leak
}
```

The missed program stores the allocation only inside an owner object. After the owner is deleted, the member allocation is unreachable, but no leak is reported:

```cpp
struct Owner {
  int *member;
  Owner() : member(new int(42)) {}
};

void test() {
  Owner *owner = new Owner;
  delete owner; // no warning for owner->member
}
```

## Runtime confirmation

An instrumented variant that replaces global `operator new` and `operator delete` with allocation counters leaves exactly one allocation outstanding after deleting the owner. The assertion succeeds when the program is executed.

## Expected result

When an owner object is destroyed without releasing allocations that are only reachable through its fields, `cplusplus.NewDeleteLeaks` should report those allocations as leaks.

## Verification

CSA version:

```text
clang version 24.0.0git (https://github.com/llvm/llvm-project.git 1cb7e838cd47ecad4050948c0c907ecb1f466ac3)
```

**Comment by steakhal** [comment]
This is actually an interesting case. This trips on a limitation, called escape-on-bind.
1) We allocate the Owner memory on heap.
2) we invoke the ctor, thus evaluate the initializer list elements, so we allocate an int on the heap.
3) We model the binding to the field of that int pointer, which triggers an escape-on-bind because we pessimistically assume that we wouldn't be able to track where the location points to anyway, which means that the value we wanted to bind escapes. The idea is that some place might legitimately release that assigned pointer somehow. This is not true in this example, but this is what's behind this.
4) We don't get to see the leak.

The escape is caused by the case (2) step here:

```c++
// A value escapes in four possible cases:
// (1) We are binding to something that is not a memory region.
// (2) We are binding to a MemRegion that does not have stack storage.
// (3) We are binding to a top-level parameter region with a non-trivial
//     destructor. We won't see the destructor during analysis, but it's there.
// (4) We are binding to a MemRegion with stack storage that the store
//     does not understand.
ProgramStateRef ExprEngine::processPointerEscapedOnBind(...)
```

Because the target of the write is on the heap somewhere, we give up.

This is not good. It's a historical shortcut to suppress reports like assigning to globals, static locals, basically anything that is not on the stack.
What we should have done is:
Only escape the value in two scenarios:
- We assign to a top-level output parameter or a location derived from it.
- We bind to the return value of the top-level function.

In these two cases the value we assign/return should legitimately escape because we don't know where this top-level function is called from. It's the call-site's responsibility to handle any resources that go there (I was tempted to say 'escape' there).

The problem is that a lot of analyzer code depends on this escape-on-bind behavior, so the semantic refinement is not so simple.
Something like this would get us half-ways, but I don't have the time to experiment with this:
```diff
diff --git a/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp b/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
index 6373e9aafe97..c1d79d8c1477 100644
--- a/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
+++ b/clang/lib/StaticAnalyzer/Core/ExprEngine.cpp
@@ -3447,9 +3447,29 @@ void ExprEngine::VisitAtomicExpr(const AtomicExpr *AE, ExplodedNode *Pred,
   getCheckerManager().runCheckersForPostStmt(Dst, AfterInvalidateSet, AE, *this);
 }
 
+static bool isDerivedFromTopLevelParamVarRegion(const MemRegion *R) {
+  if (!R)
+    return false;
+  if (const auto *SR = R->getSymbolicBase()) {
+    if (const auto *SRV = dyn_cast<SymbolRegionValue>(SR->getSymbol())) {
+
+      // Top-level parameters are unfortunately not represented by
+      // ParamVarRegions. See the comment at that class.
+      const auto *VarR = dyn_cast<VarRegion>(SRV->getRegion());
+      if (VarR && isa<ParmVarDecl>(VarR->getDecl())) {
+        return true;
+      }
+      return isDerivedFromTopLevelParamVarRegion(SRV->getRegion());
+    }
+  }
+
+  return false;
+}
+
 // A value escapes in four possible cases:
 // (1) We are binding to something that is not a memory region.
-// (2) We are binding to a MemRegion that does not have stack storage.
+// (2.a) We are binding to a MemRegion of a top-level output parameter.
+// (2.b) We are binding to a MemRegion of a top-level function.
 // (3) We are binding to a top-level parameter region with a non-trivial
 //     destructor. We won't see the destructor during analysis, but it's there.
 // (4) We are binding to a MemRegion with stack storage that the store
@@ -3459,14 +3479,24 @@ ProgramStateRef ExprEngine::processPointerEscapedOnBind(
     const StackFrame *SF, PointerEscapeKind Kind, const CallEvent *Call) {
   SmallVector<SVal, 8> Escaped;
   for (const std::pair<SVal, SVal> &LocAndVal : LocAndVals) {
-    // Cases (1) and (2).
     const MemRegion *MR = LocAndVal.first.getAsRegion();
     const MemSpaceRegion *Space = MR ? MR->getMemorySpace(State) : nullptr;
-    if (!MR || !isa<StackSpaceRegion, StaticGlobalSpaceRegion>(Space)) {
+
+    // Case (1).
+    if (!MR) {
       Escaped.push_back(LocAndVal.second);
       continue;
     }
 
+    // Case (2.a).
+    if (isDerivedFromTopLevelParamVarRegion(MR)) {
+      Escaped.push_back(LocAndVal.second);
+      continue;
+    }
+
+    // Case (2.b).
+    // TODO.
+
     // Case (3).
     if (const auto *VR = dyn_cast<VarRegion>(MR->getBaseRegion()))
       if (isa<StackArgumentsSpaceRegion>(Space) &&
@@ -3570,6 +3600,7 @@ void ExprEngine::evalBind(ExplodedNodeSet &Dst, const Stmt *StoreE,
     ProgramStateRef State = PredI->getState();
 
     // Check and record that 'Val' may escape:
+    // tmp - the escape-on-bind is called from here:
     State = processPointerEscapedOnBind(State, Location, Val, SF);
 
     if (auto AsLoc = Location.getAs<Loc>()) {
```
